### DISCUSSION TITLE
Use of Scip after license change

### DISCUSSION THREAD

**Discussion by jjhforrest** [coin-or/Cbc#573] [Discussion]
I have been trying to improve Cbc, and searching around, I discovered that the Scip license has changed from the ZIB Academic license to Apache 2.  This make it's use much more attractive in Coin.  I have added some code to Cbc to use Scip.  There is so much documentation, that I found it difficult to find everything I wanted - so I would appreciate help.  To use Scip you need to define -DCBC_TRY_SCIP in coinbrew and use -I, -L etc to point to files.  Then adding -options usescip will use scip after Cbc does work at root node.   You may want to switch off Cgl cuts.  Cbc heuristics would be good - but there I need help as to how to pass in a solution to Scip.  To encourage someone to play with the code, I put my attempt to use Scip in CbcGeneral.cpp.  You will see TODO in several places and that is where I would appreciate help.

I think this could be important.

Since I am here, I will mention two other "improvements".  In the standalone solver you can now add -optionfile filename.  This can have commands you keep using - to make it easier if you always use several commands.  So my file has lines like

* I think this works well
preprocess equalall
* switches cuts on relaxed version of model with root cuts
more2 lagrangean2
* for fast solves - this seems to do better than default
depth 5
or the file could just have those all on one line as -preprocess equalall -more2 lagrangean2 -depth 5 

This is actually a first step to try and make options more readable.

The second "improvement" is to allow more than 2147483647 nodes!

**Comment by svigerske** [comment]
To pass a solution before SCIP starts solving, better use `SCIPcreateOrigSol()` or `SCIPcreatePartialSol()` to create one, use `SCIPsetSolVals()` to set the values, and add the solution to SCIP via `SCIPaddSolFree()`.    
`SCIPsetSolVals()` should probably get your `tempvars` and `model->bestSolution()` arrays as input.   
The `SCIPcreatePartialSol()` is to give a solution that doesn't have all values set. You may need to set some options to have SCIP trying to complete the partial sol to a feasible complete one.

Maybe have a look at the code starting at https://github.com/coin-or/GAMSlinks/blob/master/src/scip/reader_gmo.c#L2127

To run a Cbc heuristic within SCIP, you could add a primal heuristic plugin to SCIP (see https://www.scipopt.org/doc-8.0.3/html/HEUR.php) that communicates with Cbc and passes solution back to SCIP.

**Reply by jjhforrest** [reply]
Thanks Stefan.  I will slowly look at your suggestions.  What I would really like is for a student to dig into all this.  OsiScpSolverInterface is one idea that comes to mind - to make moving backwards and forwards simpler.  Also a way of using Scip's preprocessing instead of CglPreprocess as Scip seems superior.  I managed to use Papilo in that way, but that adds extra complications.

Ted - yes probably a good idea to make a separate branch.

**Comment by tkralphs** [comment]
@jjhforrest Could we maybe do this in a separate branch? I would still like to make a new version out of what is currently in master. If you have the bandwidth, could we maybe try to get what is in master closer to release? I recall that there were some weird regressions with what is in master that we were working through, but it kind of got dropped. I still don't have a lot of bandwidth for this, but have a student working a lot with Cbc, so maybe we could move things forward a bit.
